cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -1354,11 +1354,17 @@ private AppFabricServiceStatus stop(Id.Program identifier, @Nullable String runI
         throw new ApplicationNotFoundException(identifier.getApplication());
       } else if (!store.programExists(identifier)) {
         throw new ProgramNotFoundException(identifier);
-      } else if (runId == null) {
-        throw new BadRequestException("Program not running");
-      } else {
+      } else if (runId != null) {
+        // Check if the program is running and is started by the Workflow
+        RunRecordMeta runRecord = store.getRun(identifier, runId);
+        if (runRecord != null && runRecord.getProperties().containsKey("workflowrunid")
+          && runRecord.getStatus().equals(ProgramRunStatus.RUNNING)) {
+          throw new BadRequestException("MapReduce program run started by Workflow cannot be stopped. " +
```

Comment:
lets add atleast the mapreduce name and run id, along with the workflow runid as well.